### ISSUE TITLE
energy/meter

### ISSUE THREAD

**Issue by gisbertg** [muppet3000/homeassistant-grott#7]
Could you take the entities for an Energy/Meter from Grott? I still have a SDM630 installed in my installation. It is read out by Grott but unfortunately not displayed by you.

 - MQTT jsonmsg: 
                         {"device": "GPG0A271VP", "time": "2023-05-16T06:50:30", "buffered": "no",
                         "values": {"datalogserial": "GPG0A271VP", "pvserial": "QCL0CGB0BC",
                         "voltage_l1": 2323, "voltage_l2": 2332, "voltage_l3": 2330, "current_l1": 4,
                         "current_l2": 8, "current_l3": 8, "act_power_l1": -334, "act_power_l2":
                         -1208, "act_power_l3": -1122, "app_power_l1": 1125, "app_power_l2": 2029,
                         "app_power_l3": 1924, "react_power_l1": 0, "react_power_l2": -1215,
                         "react_power_l3": -1246, "powerfactor_l1": -305, "powerfactor_l2": -610,
                         "powerfactor_l3": -582, "pos_rev_act_power": -2664, "pos_act_power",
                         "rev_act_power": -2664, "app_power": 5049, "react_power": -2573,
                         "powerfactor": -529, "frequency": 499, "L1-2_voltage": 4032, "L2-3_voltage":
                         4038, "L3-1_voltage": 4030, "pos_act_energy": 2931, "rev_act_energy": 5289}}
         - Grott MQTT topic used : energy/meter
         - MQTT message message sent
         - Grott Send data to PVOutput disabled 
         - Grott Send data to Influx disabled 
         - Grott extension processing disabled 

Translated with www.DeepL.com/Translator (free version)

**Comment by muppet3000** [comment]
Hi @gisbertg 

I think you're referring to this set of values from grott: https://github.com/johanmeijer/grott/blob/master/grottconf.py#L1052-L1092

I need to work out what the 'human readable' versions of all these names are, once I've done that I'll be able to map them through for you.

**Comment by muppet3000** [comment]
I've asked the developer of Grott for this information here: https://github.com/johanmeijer/grott/discussions/340

If he doesn't come back to me I'll have to try and do it a different way, perhaps if you'd be willing to share access to your growatt account I could look it up that way. But let's see what the developer comes back with first.

@shocker2 - I've noticed you have one of these meters too based on your post here: https://github.com/johanmeijer/grott/issues/339 perhaps you'd be willing to assist with the correct names for these fields?

**Comment by gisbertg** [comment]
That's exactly what they are. You can also learn more about it here.
https://github.com/johanmeijer/grott/issues/88
As shown here would be perfect I think.
https://github.com/johanmeijer/grott/issues/88#issuecomment-1013877389

Of course I don't know how many users have this setup, but there are quite a few I think.

**Comment by muppet3000** [comment]
@gisbertg  thanks that's a great start.

Can I confirm that you do get some of the information showing e.g. pvserial? You're just missing a load of the useful information e.g. all the data values.

**Comment by gisbertg** [comment]
![image](https://github.com/muppet3000/homeassistant-grott/assets/33952887/caee9997-95b3-4219-a239-dac00d286160)

32 entities are shown to me in Homeassistant.

**Comment by muppet3000** [comment]
@gisbertg what does it show on the `device` page please?
Settings -> Devices & Services -> Grott

![image](https://github.com/muppet3000/homeassistant-grott/assets/10612068/72c7f79a-c4c4-4f8c-afd4-6dee8f6f0391)

How many devices do you have?

What information is shown when you click on each device?
![image](https://github.com/muppet3000/homeassistant-grott/assets/10612068/14acb6d0-5e99-40f2-bb2c-00042ebc9d00)

**Comment by shocker2** [comment]
@muppet3000 happy to help, not sure what the ask is :) can you please provide additional info?
I do have SDM630 Modbus v3 as mettering and in HA Grott I do have two devices.
One is the invertor that we have discussed with all the values, the other is the data logger with two values, Datalog Serial and PV Serial. Not sure if there is in Grott logs anything related to the meter as I cannot see anything related to the export power.

![image](https://github.com/muppet3000/homeassistant-grott/assets/24977727/c3f62846-bd4f-45f1-b13e-815027e65912)

**Comment by gisbertg** [comment]
![image](https://github.com/muppet3000/homeassistant-grott/assets/33952887/7afe586e-a107-44e5-a732-d23fd9bd9500)
![image](https://github.com/muppet3000/homeassistant-grott/assets/33952887/2603918e-b3c7-403a-8ee2-00c9addc57ab)

**Comment by shocker2** [comment]
I have attached the log file if that helps

[grott.log](https://github.com/muppet3000/homeassistant-grott/files/11488364/grott.log)

**Comment by shocker2** [comment]
> @shocker2  - what are the two devices that you have registered in Home Assistant? Is one of them the SDM630 with just a very small amount of information e.g. pvserial and dataloggerserial?

No, one is the invertor (with all the values) and the other one is the data logger. I have identified them based on the serial number.

**Comment by muppet3000** [comment]
> 
> > @shocker2  - what are the two devices that you have registered in Home Assistant? Is one of them the SDM630 with just a very small amount of information e.g. pvserial and dataloggerserial?
> 
> No, one is the invertor (with all the values) and the other one is the data logger. I have identified them based on the serial number.

Would it be possible to get a screen grab of what is shown on each of the 'device' pages for them?

I'm sure I've seen the SDM mentioned in your log files so I'm surprised it's not showing up in HA at all.

**Comment by gisbertg** [comment]
By the way, why do I only have 1 device?

**Comment by muppet3000** [comment]
> By the way, why do I only have 1 device?

That was my next question, in the grott logs are all of your devices showing as being output to MQTT?

When you added the Grott Integration to Home Assistant did you leave it as the default '+' setting so that it includes all devices?

**Comment by shocker2** [comment]
> Would it be possible to get a screen grab of what is shown on each of the 'device' pages for them?
> 
> I'm sure I've seen the SDM mentioned in your log files so I'm surprised it's not showing up in HA at all.

![image](https://github.com/muppet3000/homeassistant-grott/assets/24977727/cdd2a3e6-edcd-4e5a-976f-43801ebad0f5)
![image](https://github.com/muppet3000/homeassistant-grott/assets/24977727/a6b18a79-7a32-44c4-946f-847e6cea6f11)
![image](https://github.com/muppet3000/homeassistant-grott/assets/24977727/f052e526-ceb8-4dbe-88ab-66dea18985ea)
![image](https://github.com/muppet3000/homeassistant-grott/assets/24977727/25fd4c47-4dea-4694-8b06-3d3f18bb8851)

**Comment by muppet3000** [comment]
Thanks @shocker2 - That second device (ending 3VK) is the SDM630 device, when I know what all the values should appear as for human values I'll be able to add them in and that will all be populated as well. I just need to get the information on what the human readable values are.
I'll try and take a look after work this evening.

**Comment by shocker2** [comment]
> Thanks @shocker2 - That second device (ending 3VK) is the SDM630 device, when I know what all the values should appear as for human values I'll be able to add them in and that will all be populated as well. I just need to get the information on what the human readable values are.
> 
> I'll try and take a look after work this evening.

Thank you! Busy Wednesday as well 😀 let me know if you need help when adding the values to identify what they are based on what I have on Shineapp.

**Comment by muppet3000** [comment]
> > Thanks @shocker2 - That second device (ending 3VK) is the SDM630 device, when I know what all the values should appear as for human values I'll be able to add them in and that will all be populated as well. I just need to get the information on what the human readable values are.
> > I'll try and take a look after work this evening.
> 
> Thank you! Busy Wednesday as well 😀 let me know if you need help when adding the values to identify what they are based on what I have on Shineapp.

That would be REALLY valuable, if you could send me a screengrab of what those values appear as on the shineapp then I should be able to work it out. 
I'm 99% sure that I've written the integration in a way that means I can change the names without losing the history of the sensor so if I don't get it perfect first time I can tweak it.

**Comment by muppet3000** [comment]
Thanks @gisbertg  - Yes it looks like they're all being published to MQTT.
The HA integration should pick them up automatically if you've used the `+` in the configuration of the integration:
![image](https://github.com/muppet3000/homeassistant-grott/assets/10612068/c785697a-fbcb-4a2a-befd-183fd11b7a7c)

**Comment by gisbertg** [comment]
![image](https://github.com/muppet3000/homeassistant-grott/assets/33952887/fdd64001-503d-4540-823a-529504957ed0)
Thats mine ...

**Comment by muppet3000** [comment]
ok in which case it should be picking everything up, they all get published to the same place in MQTT and the integration should pick them all up automatically.

I'll have to have a think about it. You could try removing and then re-adding the integration, but I don't think that would fix it.

**Comment by gisbertg** [comment]
![image](https://github.com/muppet3000/homeassistant-grott/assets/33952887/17f30b70-4215-4c28-bbf4-69c5eb4d4549)
The funny thing is that he seems to know both devices, at least he knows both serials.

**Comment by shocker2** [comment]
```
grott        |   - date-time:  2023-05-17T09:24:21
grott        |   - Grott values retrieved:
grott        |           -  datalogserial        :  the_3VK
grott        |           -  pvserial             :  the_07Y
grott        |           -  voltage_l1           :  218.7
grott        |           -  voltage_l2           :  228.6
grott        |           -  voltage_l3           :  227.3
grott        |           -  Current_l1           :  24.0
grott        |           -  Current_l2           :  14.3
grott        |           -  Current_l3           :  12.2
grott        |           -  act_power_l1         :  5270.3
grott        |           -  act_power_l2         :  3187.9
grott        |           -  act_power_l3         :  2750.0
grott        |           -  app_power_l1         :  5235.8
grott        |           -  app_power_l2         :  3256.3
grott        |           -  app_power_l3         :  2742.5
grott        |           -  react_power_l1       :  -129.1
grott        |           -  react_power_l2       :  -776.1
grott        |           -  react_power_l3       :  -378.3
grott        |           -  powerfactor_l1       :  1.0
grott        |           -  powerfactor_l2       :  1.0
grott        |           -  powerfactor_l3       :  1.0
grott        |           -  pos_rev_act_power    :  11208.2
grott        |           -  pos_act_power        :  11208.2
grott        |           -  rev_act_power        :  11208.2
grott        |           -  app_power            :  11241.4
grott        |           -  react_power          :  -1331.3
grott        |           -  powerfactor          :  1.0
grott        |           -  frequency            :  49.9
grott        |           -  L1-2_voltage         :  387.6
grott        |           -  L2-3_voltage         :  394.8
grott        |           -  L3-1_voltage         :  386.3
grott        |           -  pos_act_energy       :  22742.6
grott        |           -  rev_act_energy       :  179.7
grott        |           -  pos_act_energy_kvar  :  16.5
grott        |           -  rev_act_energy_kvar  :  6674.4
grott        |           -  app_energy_kvar      :  23878.9
grott        |           -  act_energy_kwh       :  22922.3
grott        |           -  react_energy_kvar    :  6691.0
grott        |   - MQTT jsonmsg: 
grott        |                   {"device": "the_3VK", "time": "2023-05-17T09:24:21", "buffered": "no",
grott        |                   "values": {"datalogserial": "the_3VK", "pvserial": "the_07Y",
grott        |                   "voltage_l1": 2187, "voltage_l2": 2286, "voltage_l3": 2273, "Current_l1":
grott        |                   240, "Current_l2": 143, "Current_l3": 122, "act_power_l1": 52703,
grott        |                   "act_power_l2": 31879, "act_power_l3": 27500, "app_power_l1": 52358,
grott        |                   "app_power_l2": 32563, "app_power_l3": 27425, "react_power_l1": -1291,
grott        |                   "react_power_l2": -7761, "react_power_l3": -3783, "powerfactor_l1": 999,
grott        |                   "powerfactor_l2": 971, "powerfactor_l3": 990, "pos_rev_act_power": 112082,
grott        |                   "pos_act_power": 112082, "rev_act_power": 112082, "app_power": 112414,
grott        |                   "react_power": -13313, "powerfactor": 992, "frequency": 499, "L1-2_voltage":
grott        |                   3876, "L2-3_voltage": 3948, "L3-1_voltage": 3863, "pos_act_energy": 227426,
grott        |                   "rev_act_energy": 1797, "pos_act_energy_kvar": 165, "rev_act_energy_kvar":
grott        |                   66744, "app_energy_kvar": 238789, "act_energy_kwh": 229223,
grott        |                   "react_energy_kvar": 66910}}
grott        |   - Grott MQTT topic used : energy/growatt
```

App view:

![IMG_0836](https://github.com/muppet3000/homeassistant-grott/assets/24977727/7c6abb25-4137-4a54-82b7-b38b0dd754c2)

![IMG_0837](https://github.com/muppet3000/homeassistant-grott/assets/24977727/0573450a-07f3-4928-91fd-fa1506839c3a)


![IMG_0838](https://github.com/muppet3000/homeassistant-grott/assets/24977727/f6130b0a-d0ed-4254-8644-3eb43a35147a)


![IMG_0839](https://github.com/muppet3000/homeassistant-grott/assets/24977727/a63b062d-6936-4276-a70c-b7c1cabeda57)

**Comment by muppet3000** [comment]
Thanks for all the info @shocker2 I may not be able to implement this all in one shot, but it would be good to try and get a start on it so you and @gisbertg can test and let me know what it looks like.

I'm thinking I'll start with the `act_power_l` values as they seem to be explainable.

Are these type of systems "3 Phase" systems and therefore each of the `l1`, `l2`, `l3` refer to different phases? That would make a lot of sense and I could then label them individually and also provide a sum of them all.

What do you think?

**Comment by gisbertg** [comment]
Yes exactly as you suspected.

**Comment by gisbertg** [comment]
https://www.michael-heck.net/wp-content/uploads/2019/08/SDM630-Modbus-V2-14f.pdf

Here you get, I think, all explanations to the transferred values. The manual is in German, interesting for you is page 55, which should also be in English.

**Comment by muppet3000** [comment]
Hi Guys, I had planned to sit down and work on this tonight, but something's come up and I can't. 
I'm away for the weekend so I'm unlikely to be able to pick this up again until next week, but it's top of my list.

**Comment by shocker2** [comment]
No rush, thank you for your awesome work!

**Comment by mriksman** [comment]
Hi All - Happy to help provide data if needed. I too have the SDM-ONE ([SDM120CTM ](https://www.eastroneurope.com/products/view/sdm120ct-modbus)). 

The integration sees two.
![image](https://github.com/muppet3000/homeassistant-grott/assets/20112339/37841f12-81d6-49b7-80f2-9b7fd7102c17)

But doesn't have any of the data from the meter.
![image](https://github.com/muppet3000/homeassistant-grott/assets/20112339/604e6b61-b662-4301-a349-45198d76a5a6)

```{"device": "XGD6CAxxxx", "time": "2023-05-20T12:45:59", "buffered": "no", "values": {"datalogserial": "XGD6CAxxxx", "pvserial": "HMG9CAS01Z", "voltage_l1": 2461, "voltage_l2": 0, "voltage_l3": 0, "Current_l1": 120, "Current_l2": 0, "Current_l3": 0, "act_power_l1": -27756, "act_power_l2": 0, "act_power_l3": 0, "app_power_l1": 29331, "app_power_l2": 0, "app_power_l3": 0, "react_power_l1": 8444, "react_power_l2": 0, "react_power_l3": 0, "powerfactor_l1": -957, "powerfactor_l2": 0, "powerfactor_l3": 0, "pos_rev_act_power": -27756, "pos_act_power": -27756, "rev_act_power": -27756, "app_power": 29331, "react_power": 8444, "powerfactor": -957, "frequency": 500, "L1-2_voltage": 0, "L2-3_voltage": 0, "L3-1_voltage": 0, "pos_act_energy": 810, "rev_act_energy": 601}}```

**Comment by muppet3000** [comment]
Hi Guys,

Sorry for not progressing this, I've been away on holiday and busy with family commitments, I'm planning on spending some time in the evenings on this so this is top of my list. I just need to remember where I got to with it!

Thanks for all the really useful feedback and offers for help.

**Comment by muppet3000** [comment]
I have just pushed a commit to the main branch that adds the following sensors:

- Phase1 Voltage
- Phase2 Voltage
- Phase3 Voltage
- Phase1 Current
- Phase2 Current
- Phase3 Current
- Phase1 Import from Grid Power
- Phase2 Import from Grid Power
- Phase3 Import from Grid Power
- Import from Grid Power
- Import from Grid Energy - Total
- Export to Grid Energy - Total

Please can people test and give me feedback on what's missing?
The obvious one that I think is missing is the PV generation values - does anyone know how to calculate this from the values that are returned?

**Comment by pmp02** [comment]
In theory, it should be something like:
`(voltage_l1 * Current_l1 / powerfactor_l1) + (voltage_l2 * Current_l2 / powerfactor_l2) + (voltage_l2 * Current_l2 / powerfactor_l2)`

But from my tests, this doesn't match what the Growatt app shows... so something is missing from the formula.

It would be useful maybe to create entities for all values that are coming from grott, when you have the time, so we can try out different formulas?

**Comment by muppet3000** [comment]
Very odd, well I've added the extra values in now, hopefully that gives you something to work with, if there's any extra ones I can plumb through (even with just best-guess names) let me know and you can then have a play.

**Comment by pmp02** [comment]
Thanks @muppet3000 !
My take on this would be to have all the extra values in HA, with their actual names from Grott, until we figure it out what they are. But at least we have them for comparison with Growatt App.

**Comment by shocker2** [comment]
I have updated from the main repo and I cannot see the import ones.

Below are all the values shown:

![image](https://github.com/muppet3000/homeassistant-grott/assets/24977727/7627c61e-67bb-40f4-a71e-98c17de20372)

**Comment by muppet3000** [comment]
@shocker2 that's really weird, are you sure you're running main? On the HACS page for the integration can you check that you've definitely got the main version and not the 1.0.1 release?
![image](https://github.com/muppet3000/homeassistant-grott/assets/10612068/6d829822-19ed-43f4-82e5-c8ba21dfea86)

**Comment by shocker2** [comment]
Well, indeed there is an issue. I have re-downloaded the main (selected main), restarted HA, and checked back it's still 1.0.1.

![image](https://github.com/muppet3000/homeassistant-grott/assets/24977727/abf8e3c7-8e0c-44e4-bef4-4902e3343b37)

![image](https://github.com/muppet3000/homeassistant-grott/assets/24977727/2c98d072-b9df-405e-a5e3-cba98cb974c7)

![image](https://github.com/muppet3000/homeassistant-grott/assets/24977727/457ffb96-1409-4a8f-b4d6-c54904509719)

Not sure why?

**Comment by muppet3000** [comment]
That's really weird, it shouldn't do that, that's part of HACS nothing to do with this integration, you've definitely clicked "download" right?

**Comment by mriksman** [comment]
I oddly had to do it twice, but eventually it worked and I am seeing 
![image](https://github.com/muppet3000/homeassistant-grott/assets/20112339/bfc0d8a6-99b3-472e-9212-732dc349037f)

**Comment by muppet3000** [comment]
I'll try and do a 'proper' versioned release later with all of this stuff on it maybe not close this ticket so that people that are having issues can consume it.

**Comment by shocker2** [comment]
Yes, I'm hitting the download button as this triggers the restart notification :)
Tried 3 more times and it worked, now I'm on 759caf0. No grid import yet, it might take a few minutes to update.

**Comment by muppet3000** [comment]
@shocker2 - Glad you got it installed in the end, let me know if they start appearing or not, sometimes they don't all get published at the same time I believe.

**Comment by shocker2** [comment]
Same values, nothing changed:

![image](https://github.com/muppet3000/homeassistant-grott/assets/24977727/9baacc8f-f033-4fc1-a90f-4a6cf24deb51)


![image](https://github.com/muppet3000/homeassistant-grott/assets/24977727/55d239a8-f3bc-4d5e-8fd1-8103026b0e9c)

**Comment by mriksman** [comment]
> I have just pushed a commit to the main branch that adds the following sensors:
> 
>     * Phase1 Voltage
> 
>     * Phase2 Voltage
> 
>     * Phase3 Voltage
> 
>     * Phase1 Current
> 
>     * Phase2 Current
> 
>     * Phase3 Current
> 
>     * Phase1 Import from Grid Power
> 
>     * Phase2 Import from Grid Power
> 
>     * Phase3 Import from Grid Power
> 
>     * Import from Grid Power
> 
>     * Import from Grid Energy - Total
> 
>     * Export to Grid Energy - Total
> 
> 
> Please can people test and give me feedback on what's missing? The obvious one that I think is missing is the PV generation values - does anyone know how to calculate this from the values that are returned?

You don't need 'PV Generation Values'. What you have just added is for a 'Smart Meter' - it is measuring grid line voltage/current/power, it has nothing to do with solar generation. 
![image](https://github.com/muppet3000/homeassistant-grott/assets/20112339/59088a7f-b460-44a4-8d85-66f442d942f5)
With this Smart Meter, you can see if your house (load) is importing power from the grid (solar is less than load) or exporting to grid (solar is more than load). So I now get two devices imported using your Grott - one for the PV Solar Inverter, and one for the Smart Meter
![image](https://github.com/muppet3000/homeassistant-grott/assets/20112339/0a1b665f-56d3-4499-b143-0d42eaa0293b)



Values look mostly good.
![image](https://github.com/muppet3000/homeassistant-grott/assets/20112339/6ebd5450-06b4-4cc6-9d6d-bd7275e70812)

* `voltage_l2, _l3` seem duplicate of `voltage_l1`, whilst the values in Grott seem correct (both `0`)
* `powerfactor_l#` can be interesting for people to see, and would be worth putting in. Values range from -1 to 1 (although, not in the way you think; -1 and 1 are the same value, it isn't decimal scale)
![image](https://github.com/muppet3000/homeassistant-grott/assets/20112339/18e81329-e7eb-4db1-906d-561f394846b3)
* Your 'Phase # Import from Grid' and 'Import from Grid' - what values are these? They should be the `act_power_l#` and `pos_rev_act_power`. I'd change the label to just 'Actual Power', as values can be positive (import) or negative (export).

![image](https://github.com/muppet3000/homeassistant-grott/assets/20112339/b861fcc2-c7bf-4841-a4c0-b5d4978b02b5)

* Not sure why `pos_act_power` shows a negative value when exporting. I would have thought this should be limited to `0`. Anyway, it can be ignored - just use `pos_rev_act_power`.
* You can leave out `app_power_l#` and `react_power_l#` - they aren't that important (in my opinion)

**Comment by muppet3000** [comment]
@mriksman - Thank you for this awesome feedback, that's exactly what I needed, it's so valuable!!!!!!

I'm working all day today but I'll try and implement all of these (and correct the icons as that's annoying me) tonight.

**Comment by muppet3000** [comment]
@mriksman - Thank you again for all your input here!

Feedback on all of the above, hopefully one at a time:

> voltage_l2, _l3 seem duplicate of voltage_l1, whilst the values in Grott seem correct (both 0)
- You are right, the next commit will fix that


> powerfactor_l# can be interesting for people to see, and would be worth putting in. Values range from -1 to 1 (although, not in the way you think; -1 and 1 are the same value, it isn't decimal scale)
- I'll add this in, however can you tell me what unit of measurement this is in so I can add it correctly?


> Your 'Phase # Import from Grid' and 'Import from Grid' - what values are these? They should be the act_power_l# and pos_rev_act_power. I'd change the label to just 'Actual Power', as values can be positive (import) or negative (export).
- Yes your assumption is correct, I'll change the labels so they are "Phase # Actual Power" etc


> Not sure why pos_act_power shows a negative value when exporting. I would have thought this should be limited to 0. Anyway, it can be ignored - just use pos_rev_act_power.
- I haven't mapped `pos_act_power` through so that's fine, `pos_rev_act_power` is mapped as "Import from Grid Power"


> You can leave out app_power_l# and react_power_l# - they aren't that important (in my opinion)
- I haven't mapped these through yet so maybe I won't bother unless anyone asks.

**Comment by mriksman** [comment]
> > powerfactor_l# can be interesting for people to see, and would be worth putting in. Values range from -1 to 1 (although, not in the way you think; -1 and 1 are the same value, it isn't decimal scale)
> 
>     * I'll add this in, however can you tell me what unit of measurement this is in so I can add it correctly?

It is unit-less.

**Comment by muppet3000** [comment]
> > > powerfactor_l# can be interesting for people to see, and would be worth putting in. Values range from -1 to 1 (although, not in the way you think; -1 and 1 are the same value, it isn't decimal scale)
> > 
> > 
> > ```
> > * I'll add this in, however can you tell me what unit of measurement this is in so I can add it correctly?
> > ```
> 
> It is unit-less.

Ok, when I'm next working on it, I'll add those 3 in and we'll consider this item closed.

**Comment by muppet3000** [comment]
@shocker2  - Did you get these extra values working in the end? I did a new release yesterday (v1.1.0) could you try downloading that and seeing if that works?

**Comment by muppet3000** [comment]
@mriksman - I've just pushed the powerfactor values - could you please pull the `main` branch at some point and check that they're correct.

**Comment by shocker2** [comment]
> @shocker2 - Did you get these extra values working in the end? I did a new release yesterday (v1.1.0) could you try downloading that and seeing if that works?

Inverter is now updated, I'll check the values tomorrow as today it's cloudy and already late.

![image](https://github.com/muppet3000/homeassistant-grott/assets/24977727/d76795f7-4bec-45f9-aa66-0c78d64e9740)

**Comment by muppet3000** [comment]
> > @shocker2 - Did you get these extra values working in the end? I did a new release yesterday (v1.1.0) could you try downloading that and seeing if that works?
> 
> Inverter is now updated, I'll check the values tomorrow as today it's cloudy and already late.
> 
> ![image](https://github.com/muppet3000/homeassistant-grott/assets/24977727/d76795f7-4bec-45f9-aa66-0c78d64e9740)
> 

That looks like the new values are now displaying so providing I've mapped them to the correct values it should all be good.

**Comment by mriksman** [comment]
> @mriksman - I've just pushed the powerfactor values - could you please pull the `main` branch at some point and check that they're correct.

You have the ones for the 3 phases, but you are missing the combined one (the most important one) `powerfactor`.

**Comment by muppet3000** [comment]
> > @mriksman - I've just pushed the powerfactor values - could you please pull the `main` branch at some point and check that they're correct.
> 
> You have the ones for the 3 phases, but you are missing the combined one (the most important one) `powerfactor`.
> 
> And I know some vendors use `-0 -> -100/100 -> 0`, it should really be `-0 -> -1/1 -> 0`. Could you divide these values by 100? This should be displaying `-0.793`. Which means `0.793 lagging`. If it was positive it would be `0.793 leading`. ![image](https://user-images.githubusercontent.com/20112339/251931067-43c699d7-bec4-40ef-af6f-52a0398b040d.png)
> 
> You can see Shelley even do it as 0-100%. Which is not right at all.. ![image](https://user-images.githubusercontent.com/20112339/251935352-7e34ed76-a592-4b1e-bbb6-7f223d629e5d.png) And see what happens when it flicks between lagging (minus) to leading (positive)? When the range is `-100 -> 0 -> 100`, you see this happen. ![image](https://user-images.githubusercontent.com/20112339/251935710-16a1e53c-4350-4c37-84eb-f38359dbd47b.png)
> 
> Personally, if I was doing it, I would have `abs(powerfactor)` and then another text entity that says `leading/lagging` depending if it is positive/negative.

@mriksman  - I've pushed a new change to add in the `powerfactor` value and also show the values between -1 and +1, can you please confirm I've done this correctly from your end.

Once I've got confirmation that's correct I'll work out if there's a nice way of adding `leading` or `lagging` but I'm not making any promises as I'd need to add it for all of the phases and the total one.

**Comment by mriksman** [comment]
Yup, can see the new `powerfactor` value and they have been divided by 100. 

If you do use `leading/lagging` remember apply `abs` to the power factor values.

**Comment by muppet3000** [comment]
Thanks for confirming, I'll attempt to clean up some of the units of measurement as well before I do a release that includes this.
What I don't want to do is turn a numeric value into a string, it makes it difficult to use in automations etc. with variables. What I may do is add a 'calculated' value that adds it as a string purely for convenience.
So you'd have all the ones we currently have plus:
- Phase1 Power factor (absolute)
- Phase2 Power factor (absolute)
- Phase3 Power factor (absolute)
- Power factor (absolute)

They'd be strings with something like: `0.79 leading` or `0.21 lagging`
